File: velhia.api/src/entities/multiAgentSystem/agent.ts
import IAgent from "./iAgent"
import Memory from "./memory"

export default class Agent implements IAgent{
  public readonly progenitor: string
  public readonly birth: Date
  public becomeLeader?: Date
  public death?: Date
  public life: number
  public memory: Memory[]
  public matchsAsLearner: number
  public matchsAsLeader: number
  public victories: number
  public defeats: number
  public draw: number

  public constructor (
    _progenitor: string,
    _birth: Date,
    _becomeLeader: Date,
    _death: Date,
    _life: number,
    _memory: Memory[],
    _matchsAsLearner: number,
    _matchsAsLeader: number,
    _victories: number,
    _defeats: number,
    _draw: number
  ) {
    this.progenitor = _progenitor
    this.birth = _birth
    this.becomeLeader = _becomeLeader
    this.death = _death
    this.life = _life
    this.memory = _memory
    this.matchsAsLearner = _matchsAsLearner
    this.matchsAsLeader = _matchsAsLeader
    this.victories = _victories
    this.defeats = _defeats
    this.draw = _draw
    Object.freeze(this)
  }
}

File: velhia.game/src/entities/agent.py
class Agent:
    """
    Agent Class
    :param obj: `IAgent` Agent object saved in MongoDB
    :param my_char: `list` List with its number and symbol ['X', 1]
    """

    def __init__(self, obj, my_char):
        self.id = obj['_id']
        self.birth = obj['birth']
        self.progenitor = obj['progenitor']
        self.becomeLeader = obj['becomeLeader'] if 'becomeLeader' in obj else ''
        self.death = obj['death'] if 'death' in obj else ''
        self.life = obj['life']
        self.memory = obj['memory']
        self.matchsAsLearner = obj['matchsAsLearner']
        self.matchsAsLeader = obj['matchsAsLeader']
        self.victories = obj['victories']
        self.defeats = obj['defeats']
        self.draw = obj['draw']
        self.char = my_char
